### PR TITLE
Dpdk backend : Support for Direct counter in .spec and context json

### PR DESCRIPTION
**Pull request by usha1830**
This PR adds support for DirectCounter extern in .spec file and context json output files for dpdk compiler
In the spec file, a new regarray is declared for direct counter extern with size equal to the owner table size + 1 for regular tables and table_size *4 + 1 for learner tables.

For the count method of direct counter, the following instructions should be emitted.
        entryid m.table_entry_index
        regadd per_prefix_pkt_bytes_count_0_packets m.table_entry_index 1

In case of PACKETS_AND_BYTES counter, two regarrays should be emitted with the counter name suffixed with "_packets" and "_bytes"
Similarly, the count method also should have "regadd" instruction for packets and bytes counter.

- Error should be issued when direct counter property coexist with implementation property.

For context json, the implementation should follow indirect counter

### CODE REVIEW

File: backends/dpdk/backend.cpp [hunk 1]
```diff
@@ -68,6 +68,8 @@ void DpdkBackend::convert(const IR::ToplevelBlock *tlb) {
         new ElimHeaderCopy(typeMap),
         new P4::TypeChecking(refMap, typeMap),
         new P4::RemoveAllUnusedDeclarations(refMap),
+        new CollectDirectCounterMeter(refMap, typeMap, &structure),
+//        new CollectDirectCounterMeter(),
```
Comment (psivanup) [inline backends/dpdk/backend.cpp]: shall be removed?

File: backends/dpdk/dpdkArch.cpp [hunk 2]
```diff
@@ -727,11 +727,15 @@ const IR::Node *InjectJumboStruct::preorder(IR::Type_Struct *s) {
     return s;
 }
 
-const IR::Node *InjectOutputPortMetadataField::preorder(IR::Type_Struct *s) {
-    if (structure->isPNA() && s->name.name == structure->local_metadata_type) {
-        s->fields.push_back(new IR::StructField(
-            IR::ID(PnaMainOutputMetadataOutputPortName), IR::Type_Bits::get(32)));
-        LOG3("Metadata structure after injecting output port:" << std::endl << s);
+const IR::Node *InjectFixedMetadataField::preorder(IR::Type_Struct *s) {
+    if  (s->name.name == structure->local_metadata_type) {
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.cpp:722]: can compare originalName instead of name?

File: backends/dpdk/dpdkArch.cpp [hunk 3]
```diff
@@ -1798,6 +1802,7 @@ getExternInstanceFromProperty(const IR::P4Table* table,
     }
 
     auto expr = property->value->to<IR::ExpressionValue>()->expression;
+    externName = expr->to<IR::PathExpression>()->path->name.name;
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.cpp:1806]: originalName?

File: backends/dpdk/dpdkArch.h [hunk 4]
```diff
@@ -670,6 +670,11 @@ class CollectExternDeclaration : public Inspector {
                     ::error(ErrorType::ERR_EXPECTED,
                             "%1%: expected number of counters and type of counter as arguments", d);
                 }
+            } else if (externTypeName == "DirectCounter") {
+                if (d->arguments->size() != 1) {
+                    ::error(ErrorType::ERR_EXPECTED,
+                            "%1%: expected type of counter as arguments", d);
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.h]: as arguments -> as only argument??

File: backends/dpdk/dpdkArch.h [hunk 5]
```diff
@@ -920,6 +925,22 @@ class ConvertActionSelectorAndProfile : public PassManager {
...
 };
 
+/* Collect size information from the owner table for direct counter and meter extern objects */
+class CollectDirectCounterMeter : public Inspector {
+    P4::ReferenceMap* refMap;
+    P4::TypeMap* typeMap;
+    DpdkProgramStructure* structure;
+ public:
+    static ordered_map<cstring, int> directMeterCounterSizeMap;
+//    CollectDirectCounterMeter() {setName("CollectDirectCounterMeter");}
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.h]: shall be removed?

File: backends/dpdk/dpdkArch.h [hunk 6]
```diff
@@ -920,6 +925,22 @@ class ConvertActionSelectorAndProfile : public PassManager {
...
+    CollectDirectCounterMeter(P4::ReferenceMap *refMap, P4::TypeMap *typeMap,
+                DpdkProgramStructure* structure) :
+    refMap(refMap), typeMap(typeMap), structure(structure) {}
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.h:994]: setName required here?

File: backends/dpdk/dpdkArch.h [hunk 7]
```diff
@@ -920,6 +925,22 @@ class ConvertActionSelectorAndProfile : public PassManager {
...
+
+    int getTableSize(const IR::P4Table * tbl);
+    void postorder(const IR::P4Table* t) override;
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.h]: these functions shall be declared private.

File: backends/dpdk/dpdkHelpers.cpp [hunk 8]
```diff
@@ -1030,6 +1030,48 @@ bool ConvertStatementToDpdk::preorder(const IR::MethodCallStatement *s) {
...
+                    } else {
+                         if (value == 1)
+                             add_instr(new IR::DpdkCounterCountStatement(counter, metaIndex,
+                                                                               incr));
+                         else
+                             add_instr(new IR::DpdkCounterCountStatement(counter, metaIndex));
+                    }
+                }
+            } else {
+                BUG("Direct Counter function not implemented");
```
Comment (psivanup) [inline backends/dpdk/dpdkHelpers.cpp]: can include the method name in the message?

**Review by psivanup — CHANGES_REQUESTED** [review]
Please check the comments.

File: backends/dpdk/dpdkArch.cpp (hunk 2 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp:722]: In many cases, originalName is also modified by the frontend. Have faced issues in the passt. We rely on the internal name which is present as name.name.

File: backends/dpdk/dpdkArch.cpp (hunk 3 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp:1806]: Spec file uses the internal name instead of originalName

File: backends/dpdk/dpdkArch.h (hunk 4 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.h]: updated

**Review by psivanup — APPROVED** [review]
Looks fine.

File: backends/dpdk/dpdkArch.h (hunk 6 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.h:994]: Updated

File: backends/dpdk/dpdkHelpers.cpp (hunk 8 above)
Comment (usha1830) [reply backends/dpdk/dpdkHelpers.cpp]: updated

File: backends/dpdk/dpdkArch.h (hunk 5 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.h]: removed

**Review by psivanup — APPROVED** [review]
Looks fine.

**Review by github-sajan — APPROVED** [review]
Code looks fine to me

**Comment by usha1830** [comment]
Closing as upstreamed

**Pull request closed without merging**